sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala

```diff
@@ -95,11 +100,20 @@ private[spark] class HiveExternalCatalog(conf: SparkConf, hadoopConf: Configurat
   }
 
   /**
-   * Run some code involving `client` in a [[synchronized]] block and wrap certain
+   * Run some code involving `client` in a [[ReadWriteLock]] and wrap certain
    * exceptions thrown in the process in [[AnalysisException]].
+   *
+   * @param db to specify the place of the operation act on.
+   * @param writeLock to specify it is a write lock.
    */
-  private def withClient[T](body: => T): T = synchronized {
+  private def withClient[T](db: String = "", writeLock: Boolean = false) (body: => T): T = {
```

Comment:
can you refer to some hive documents? I'm not sure it's safe to assume that we can lock on the database-level. I thought the hive client itself is not thread-safe.

Reply:
@cloud-fan The hive client is not thread-safe.
1. If there are two threads are creating two tables in the same database. For example, thread-1 is creating db1.table1 and thread-2 is creating db1.table2. This db-level lock in `withClient` will block one thread until the other one finished.
2. If there are two threads are creating two tables in the different database. For example, thread-1 is creating db1.table1 and thread-2 is creating db2.table2.  There operations could be executed concurrently since the hive clients are thread-local. The client in thread-1 is not the same instance with the client in thread-2:
```scala
  private def client: Hive = {
    // get the Hive and set to thread local
    val c = Hive.get(conf)
    Hive.set(c)
    c
  }
```

Reply:
Interesting. If it's thread-local, then `HiveExternalCatalog.client` should be thread-safe? Or these hive clients that in different threads still share something?

Reply:
You are right, you remind me to check the diff between our code and spark3 again. We changed many objects in `HiveExternalCatalog.client` non-shared. For example,
In spark3
```scala
  val state: SessionState = ...

  override val userName = UserGroupInformation.getCurrentUser.getShortUserName
```
and our code
```scala
  // get the state from thread-local
  def state: SessionState = ...

  // This should be def instead of val. Since it may be invoked in different ugi
  // in thrift server, we need re-evaluate it every time it is invoked, and internally
  // it obtains the user name via the current ugi.
  private def userName = conf.getUser
```

Em, let me remark it to WIP.

Reply:
We use the spark thrift server only, so the `HiveExternalCatalog.client.state` always thread-local. But for the open source spark, it may be a little bit tricky to change.